Teach EagerSpecializer how to use the new form of the @_specialize attribute

swiftix opened the pull request:
Teach EagerSpecializer how to use the new form of the @_specialize attribute
    
In addition to supporting the creation of full specializations, the EagerSpecializer changes contain some code for generating the layout-constrained partial specializations as well.

swiftix committed:
Generic specializer API changes to prepare for the new @_specialize attribute and partial specializations.

swiftix committed:
Teach EagerSpecializer how to use the new form of the @_specialize attribute

In addition to supporting the creation of full specializations, the EagerSpecializer changes contain some code for generating the layout-constrained partial specializations as well.

swiftix commented:
@swift-ci please smoke test

slavapestov commented on lib/SILOptimizer/IPO/EagerSpecializer.cpp:
```diff
@@ -193,22 +199,48 @@ emitApplyWithRethrow(SILBuilder &Builder,
-  assert(!CanSILFuncTy->isPolymorphic());
+  auto CalleeSubstFnTy = CanSILFuncTy;
+  ArrayRef<Substitution> Subs;
+  //assert(!CanSILFuncTy->isPolymorphic());
```
Remove this line?

slavapestov commented on lib/SILOptimizer/IPO/EagerSpecializer.cpp:
```diff
@@ -193,22 +199,48 @@ emitApplyWithRethrow(SILBuilder &Builder,
-    assert(!CanSILFuncTy->isPolymorphic());
-    return Builder.createApply(CalleeFunc->getLocation(), FuncRefInst, CallArgs,
-                               false);
+    //assert(!CanSILFuncTy->isPolymorphic());
```
Ditto

swiftix committed:
Add mangling support for partial specializations.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test

slavapestov commented:
Sorry, the merge conflicts are from the `ArrayRef<Substitution> => SubstitutionList` renaming. They should be completely mechanical to resolve.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test

swiftix committed:
Teach SIL type lowering about trivial, but address only types.

swiftix committed:
Teach IRGen how to work with archetypes of fixed sizes.

This allows IRGen to use direct loads and stores instead of calling value witness tables.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@slavapestov OK, so this is supposed to be a working version of the @_specialize. I implemented the code generation for the dynamic dispatch and IRGen changes. Tests are updated to test these new features.

The function `ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF, ArrayRef<Requirement> Requirements) ` in `Generics.cpp` may need to be improved later to use your conformance lookup fixes.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test

swiftix committed:
Fix the test to work on objc-interop enabled and disabled platforms.

swiftix commented:
@swift-ci please smoke test Linux

swiftix commented:
@swift-ci please smoke test Linux

swiftix commented:
@swift-ci please smoke test Linux

swiftix commented:
@swift-ci please smoke test Linux

swiftix commented:
@swift-ci please smoke test OS X

swiftix commented:
@swift-ci please smoke test OS X

swiftix commented:
@slavapestov Anything to be fixed before merging?

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -127,6 +135,109 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+      FnTy->getOptionalErrorResult(), M.getASTContext());
+
+  // This is an interface type. It should not have any archetypes.
+  if (NewFnTy->hasArchetype()) {
```
Remove these three lines

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+      assert(FirstType && SecondType);
+
+      bool isFirstTypeNonConcrete =
+          FirstType->hasArchetype() || FirstType->hasTypeParameter();
```
hasArchetype should always be false -- can you assert?

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+      bool isSecondTypeNonConcrete =
+          SecondType->hasArchetype() || SecondType->hasTypeParameter();
+      // Only one of the types should be concrete.
+      assert((isFirstTypeNonConcrete ^ isSecondTypeNonConcrete) &&
```
I would find != more readable than ^ here

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+
+  for (auto &Req : Requirements) {
+    if (Req.getKind() == RequirementKind::SameType) {
+      auto CallerArchetype = dyn_cast<SubstitutableType>(
```
dyn_cast can return null; either use cast or handle the null case

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+              ->getCanonicalType());
+      // Remember that a given generic parameter is mapped
+      // to a concrete type.
+      CallerArchetypeToInterfaceMap.addSubstitution(
```
addSubstitution() is going to be private. I just refactored a dozen usages of it, please don't add new ones. Substitution maps should be constructed by using GenericSignature::getSubstitutionMap() or GenericEnvironment::getSubstitutionMap()

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+  for (auto Sub : ForwardingSubs) {
+    auto ClonerSub = Sub.subst(
+        SM, QuerySubstitutionMap{ClonerArchetypeToConcreteMap},
+        LookUpConformanceInModule(SM));
```
Add a FIXME comment to clean up the module conformance lookup (for me :))

slavapestov commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+  // Substitute into forwarding substitutions to get the cloner substitutions
+  // and substitutions to be used by the caller for calling the specialized
+  // function.
+  SmallVector<Substitution, 4> ClonerSubsList;
```
Instead, construct these arrays by calling GenericSignature::getSubstitutions() or GenericEnvironment::getSubstitutions(). Otherwise you're relying on them having entries in the same order as the forwarding substitutions.

swiftix committed:
Code cleanups based @slavapestov's review

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

slavapestov commented:
Ok, I think this is good to go now.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test
